pdns/validate.cc

```diff
@@ -661,18 +674,28 @@ dState getDenial(const cspmap_t &validrrsets, const DNSName& qname, const uint16
         nsec3Seen = true;
 
         LOG("\tquery hash: "<<toBase32Hex(h)<<endl);
-        string beginHash=fromBase32Hex(v.first.first.getRawLabels()[0]);
+        string beginHash = fromBase32Hex(hashedOwner.getRawLabels()[0]);
 
         // If the name exists, check if the qtype is denied
         if (beginHash == h) {
 
+          /* The NSEC is either a delegation one, from the parent zone, and
+           * must have the NS bit set but not the SOA one, or a regular NSEC
+           * either at apex (signer == owner) or with the SOA or NS bits clear.
+           */
+          const bool notApex = signer.countLabels() < qname.countLabels();
+          if (notApex && nsec3->isSet(QType::NS) && nsec3->isSet(QType::SOA)) {
+            LOG("However, that NSEC3 is not at the apex and has both the NS and the SOA bits set!"<<endl);
+            continue;
+          }
+
           /* RFC 6840 section 4.1 "Clarifications on Nonexistence Proofs":
              Ancestor delegation NSEC or NSEC3 RRs MUST NOT be used to assume
              nonexistence of any RRs below that zone cut, which include all RRs at
              that (original) owner name other than DS RRs, and all RRs below that
              owner name regardless of type.
           */
-          if (qtype != QType::DS && isNSEC3AncestorDelegation(signer, v.first.first, nsec3)) {
+          if (qtype != QType::DS && isNSEC3AncestorDelegation(signer, qname, nsec3)) {
```

Comment:
In addition to the new check above this was also an already existing bug?
It worries me a b it that this test is similar, but not equal to the NSEC case. If possible it would be good to make them more alike.

Reply:
It's more correct that way but I'm not sure there is an actual bug. If the record comes from the parent zone (DS) the hashed and unhashed names will have more labels than the signer name, by definition. If it comes from the child zone, the hashed and unhashed names will have at least as many labels than the signer name.

Reply:
As for making that case more similar to the NSEC case, it's complicated as the depth of the zone is lost with NSEC3 until the `closest encloser` has been found.